src/operator/numpy/random/np_gamma_op.h

```diff
@@ -0,0 +1,422 @@
+  using namespace mxnet_op;
+  const NumpyGammaParam &param = nnvm::get<NumpyGammaParam>(attrs.parsed);
+  CHECK_EQ(outputs.size(), 1);
+  Stream<xpu> *s = ctx.get_stream<xpu>();
+  // Generate base random number.
+  Random<xpu, FType> *prnd = ctx.requested[0].get_random<xpu, FType>(s);
+  index_t output_len = outputs[0].Size();
+  Tensor<xpu, 1, FType> random_tensor =
+      ctx.requested[1].get_space_typed<xpu, 1, FType>(Shape1(output_len * 2 * M + 1), s);
+  Tensor<xpu, 1, FType> uniform_tensor = random_tensor.Slice(0, output_len * M);
+  Tensor<xpu, 1, FType> normal_tensor = random_tensor.Slice(output_len * M, output_len * 2 * M);
+  prnd->SampleUniform(&uniform_tensor, 0, 1);
+  prnd->SampleGaussian(&normal_tensor, 0, 1);
+  mxnet::TShape new_lshape, new_hshape, new_oshape;
+  FType failure_indicator = 1.0;
+  Tensor<xpu, 1, FType> failure_indic_workspace =
+      random_tensor.Slice(output_len * 2 * M, output_len * 2 * M + 1);
+  FType *failure_indicator_device = failure_indic_workspace.dptr_;
+  // [scalar scalar] case
+  if (inputs.size() == 0U) {
+    MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, OType, {
+      Kernel<gamma_two_scalar_kernel<OType, FType>, xpu>::Launch(
+          s, outputs[0].Size(), param.shape.value(), param.scale.value(),
+          uniform_tensor.dptr_, normal_tensor.dptr_, outputs[0].dptr<OType>());
+      Kernel<CheckSuccessKernel<OType, FType>, xpu>::Launch(
+          s, outputs[0].Size(), outputs[0].dptr<OType>(),
+          failure_indicator_device);
+      _copy<xpu>(s, &failure_indicator, failure_indicator_device);
+      while (1) {
+        if (failure_indicator >= 0) {
+          break;
+        } else {
+          prnd->SampleUniform(&uniform_tensor, 0, 1);
+          prnd->SampleGaussian(&normal_tensor, 0, 1);
+          Kernel<gamma_two_scalar_kernel_r<OType, FType>, xpu>::Launch(
+              s, outputs[0].Size(), param.shape.value(), param.scale.value(),
+              uniform_tensor.dptr_, normal_tensor.dptr_,
+              outputs[0].dptr<OType>(), failure_indicator_device);
+          failure_indicator = 1.0;
+          Kernel<CheckSuccessKernel<OType, FType>, xpu>::Launch(
+              s, outputs[0].Size(), outputs[0].dptr<OType>(),
+              failure_indicator_device);
+          _copy<xpu>(s, &failure_indicator, failure_indicator_device);
+        }
+      }
+    });
+  } else if (inputs.size() == 1U) {
+    // [scalar tensor], [tensor scalar] case
+    int ndim = FillShape(inputs[0].shape_, inputs[0].shape_, outputs[0].shape_,
+                         &new_lshape, &new_lshape, &new_oshape);
+    int scalar_pos;
+    float scalar_value;
+    // int type_flag = param.t;
+    if (param.shape.has_value()) {
+      scalar_pos = 0;
+      scalar_value = param.shape.value();
+    } else {
+      scalar_pos = 1;
+      scalar_value = param.scale.value();
+    }
+    MSHADOW_TYPE_SWITCH(inputs[0].type_flag_, IType, {
+      MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, OType, {
```

Comment:
Seems like your output type can only be floating-point numbers, so you can use `MSHADOW_REAL_TYPE_SWITCH` instead.